### ISSUE TITLE
Orb Install step failed

### ISSUE THREAD

**Issue by kmeelu** [CircleCI-Public/aws-sam-serverless-orb#20]
### Orb version
1.0.2

### What happened
![Screen Shot 2020-08-31 at 2 20 39 PM](https://user-images.githubusercontent.com/3261418/91769942-27e20600-eb95-11ea-971d-44057eafc571.png)

CircleCI Step to [install SAM CLI](https://circleci.com/orbs/registry/orb/circleci/aws-serverless#commands-install) failed. Here is the log: https://gist.github.com/kmeelu/b6483118368abe63de8a8fd6b8f1c74e

I'm not sure whether the follow up is with this Orb or SAM CLI, but pinging here in case you're aware.

### Expected behavior

SAM CLI installs as expected, allowing us to deploy lambdas.

**Comment by jazott3** [comment]
I think this issue is related. I've done a cursory pass over the upstream repositories, but haven't found any obvious signs of what broke: https://github.com/aws/homebrew-tap/issues/130

**Comment by DanielFrank** [comment]
> I think this issue is related

I ran into that issue in CircleCI. I reported it there since I assumed AWS's homebrew-tap would have an idea? (Also they were helpful when their upgrading to Python v3.8 broke things...)

**Comment by threadstonesecure** [comment]
I am facing same issue mentioned above with **circleci/aws-sam-serverless@1.0.1**

**Comment by syedhaseebahmed** [comment]
I am facing same issue mentioned above with circleci/aws-sam-serverless@1.0.1

**Comment by jazott3** [comment]
I had some time to take a closer look this morning. It looks like the "Install Linuxbrew" step is no longer working as expected. In our builds, I'm seeing the following line:

```
Warning: /home/linuxbrew/.linuxbrew/bin is not in your PATH.
```

If I re-run the job with SSH and issue:
```
/home/linuxbrew/.linuxbrew/bin/brew tap aws/tab
/home/linuxbrew/.linuxbrew/bin/brew install aws-sam-cli
/home/linuxbrew/.linuxbrew/bin/sam --version
```

I get the expected output: `SAM CLI, version 1.1.0`

I'm still not sure what changed, but I believe fixing the path should rectify this issue

Edit: I inspected some of our earlier builds, and they throw the same warning, so it's possible there's some other issue with the path that needs fixing.

**Comment by DanielFrank** [comment]
As an experiment, I noticed that Linuxbrew is also complaining that we should switch to Homebrew (Specifically `https://raw.githubusercontent.com/Homebrew/install/master/install.sh`. I tried that and it didn't work by itself. (Same issue). Trying @jazott3 's fix now.

Minor note that after installing Homebrew/LinuxBrew, we pipe to `bash /home/linuxbrew/.linuxbrew/bin/brew shellenv >> $BASH_ENV` which maybe was supposed to deal with getting LinuxBrew bin in our path?

**Comment by DanielFrank** [comment]
> I'm still not sure what changed, but I believe fixing the path should rectify this issue

I tried the path fixes with both Linuxbrew install and Homebrew install and it did not work. I threw in a `ls /home/linuxbrew/.linuxbrew/bin/` just to see if `sam` is even in the directory and got:
```
brew	 bzip2recover  infocmp	  lzma		    sqlite3  xzcat
bunzip2  bzless        infotocap  lzmadec	    tabs     xzcmp
bzcat	 bzmore        lzcat	  lzmainfo	    tic      xzdec
bzcmp	 c_rehash      lzcmp	  lzmore	    toe      xzdiff
bzdiff	 captoinfo     lzdiff	  ncurses6-config   tput     xzegrep
bzegrep  clear	       lzegrep	  ncursesw6-config  tset     xzfgrep
bzfgrep  gdbm_dump     lzfgrep	  openssl	    unlzma   xzgrep
bzgrep	 gdbm_load     lzgrep	  patchelf	    unxz     xzless
bzip2	 gdbmtool      lzless	  reset		    xz	     xzmore
```

Note that there's no sam.

I also, just for fun, ran `echo $PATH` and got:
`/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/circleci/.local/bin:/home/circleci/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin` so I don't think path is the issue?

I'm consistently getting 
```
Reinstalling 1 broken dependent from source:
aws/tap/aws-sam-cli
```
Did this appear in your experiment?

**UPDATE** Note that I didn't ssh; I just ran these as `command`s. I assume by SSH, you mean [this](https://circleci.com/docs/2.0/ssh-access-jobs/)?

**Comment by jpalmonacid** [comment]
I've copied the https://circleci.com/orbs/registry/orb/circleci/aws-sam-serverless#commands-install as a new command in my project's `config.yml` file  and added the following lines in the next step:
```shell
- run:
  command: |
    brew tap aws/tap
    brew install aws-sam-cli
    ln -s /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0.reinstall /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0
    echo export PATH="/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0/bin:${PATH}" >> $BASH_ENV
    export PATH="/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0/bin:${PATH}"
    sam --version
  name: Install the AWS SAM CLI
```
This would be a workround, not the proper solution. I hope it helps you temporarily until it gets fixed.

**Comment by jazott3** [comment]
@DanielFrank 
> I'm consistently getting
`Reinstalling 1 broken dependent from source:`
`aws/tap/aws-sam-cli`

I do see that in our build log.

> I assume by SSH, you mean this?

Yes, that's correct. 

> Minor note that after installing Homebrew/LinuxBrew, we pipe to bash /home/linuxbrew/.linuxbrew/bin/brew shellenv >> $BASH_ENV which maybe was supposed to deal with getting LinuxBrew bin in our path?

I think this is correct too.

I'll run a few more tests later today. I didn't see `/home/linuxbrew/.linuxbrew/bin` (or any Linuxbrew paths) when I ran `echo $PATH` from inside the ssh session, but I wouldn't be surprised if the session is using a different profile than the run commands.

**Comment by DanielFrank** [comment]
> This would be a workround, not the proper solution

An alternative workaround to @jpalmonacid 's solution (which works fine!) is:
```
        command: |
          brew tap aws/tap
          brew install aws-sam-cli
          ln -s /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0.reinstall /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0
          ln -s ../Cellar/aws-sam-cli/1.1.0/bin/sam /home/linuxbrew/.linuxbrew/bin
          sam --version
```

Or skip soft-linking `1.10.reinstall` to `1.1.0` and just do 
```
 ln -s ../Cellar/aws-sam-cli/1.1.0.reinstall/bin/sam /home/linuxbrew/.linuxbrew/bin
```

Note this workaround will break when/if things get fixed. The issues seem to be the reinstall and that it's not soft-linking to the Linuxbrew bin.

**Comment by gmemstr** [comment]
Hi all, thanks for bringing this to our attention! Digging a little deeper on our end - very curious about what has changed.

**Comment by DanielFrank** [comment]
Possibly the Homebrew update is involved? See [here](https://github.com/aws/homebrew-tap/issues/132)

**Comment by lightpriest** [comment]
We've stumbled upon this as well, as a simple workaround you can call `brew install ..` twice.

I guess some kind of idempotency handling triggered it to place the executable at the correct location.

**Comment by BeyondEvil** [comment]
We hit this issue as well. Instead of using brew, we simply installed it using pip:

```sh
pip install aws-sam-cli==1.1.0
```

🤷‍♂️

**Comment by KyleTryon** [comment]
Hey all thanks for bringing this to our attention.
I'm looking into this as well as other updates to the orb. 

@BeyondEvil thank you for mentioning the pip installation, that does not appear to be in AWS docs. I am not sure if we want to limit folks to a Python environment necessarily, but the jobs currently run within the machine executor for Docker support, which has Python. So I am considering that as a solution. 

I've also taken a little back channel to see if we can get help with https://github.com/aws/homebrew-tap/issues/132, thank you @DanielFrank

**Comment by DanielFrank** [comment]
Question for @KyleTryon : Should I close https://github.com/aws/homebrew-tap/issues/130 since it's a duplicate of https://github.com/aws/homebrew-tap/issues/132 ? I don't know if having two issues is better or worse...

**Comment by Kellym-Kainos** [comment]
> > This would be a workround, not the proper solution
> 
> An alternative workaround to @jpalmonacid 's solution (which works fine!) is:
> 
> ```
>         command: |
>           brew tap aws/tap
>           brew install aws-sam-cli
>           ln -s /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0.reinstall /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0
>           ln -s ../Cellar/aws-sam-cli/1.1.0/bin/sam /home/linuxbrew/.linuxbrew/bin
>           sam --version
> ```
> 
> Or skip soft-linking `1.10.reinstall` to `1.1.0` and just do
> 
> ```
>  ln -s ../Cellar/aws-sam-cli/1.1.0.reinstall/bin/sam /home/linuxbrew/.linuxbrew/bin
> ```
> 
> Note this workaround will break when/if things get fixed. The issues seem to be the reinstall and that it's not soft-linking to the Linuxbrew bin.

Anyone still getting `sam: command not found` with these workarounds?

**Comment by jpalmonacid** [comment]
> > > This would be a workround, not the proper solution
> > 
> > 
> > An alternative workaround to @jpalmonacid 's solution (which works fine!) is:
> > ```
> >         command: |
> >           brew tap aws/tap
> >           brew install aws-sam-cli
> >           ln -s /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0.reinstall /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.1.0
> >           ln -s ../Cellar/aws-sam-cli/1.1.0/bin/sam /home/linuxbrew/.linuxbrew/bin
> >           sam --version
> > ```
> > 
> > 
> > Or skip soft-linking `1.10.reinstall` to `1.1.0` and just do
> > ```
> >  ln -s ../Cellar/aws-sam-cli/1.1.0.reinstall/bin/sam /home/linuxbrew/.linuxbrew/bin
> > ```
> > 
> > 
> > Note this workaround will break when/if things get fixed. The issues seem to be the reinstall and that it's not soft-linking to the Linuxbrew bin.
> 
> Anyone still getting `sam: command not found` with these workarounds?

@Kellym-Kainos That's because a new release of sam has come out. You may, as I did, follow what @lightpriest suggested and it'll work fine,whatever the version of `sam` is. Install sam twice.

> We've stumbled upon this as well, as a simple workaround you can call `brew install ..` twice.
> 
> I guess some kind of idempotency handling triggered it to place the executable at the correct location.

**Comment by DanielFrank** [comment]
I've not had a chance to check but apparently it works with Homebrew 2.5
See [here](https://github.com/aws/homebrew-tap/issues/130#issuecomment-689159583)

Will check but giving someone else the chance to check first...

**UPDATE** Getting the same error.

**Comment by jfuss** [comment]
Has anyone cut a Github issue to Homebrew/Linuxbrew directly? SAM CLI 1.1.0 was released on Aug 12 and 1.2.0 was released Sept 3. Considering there was no changes to the Homebrew tap, this smells like a regression/bug within Homebrew/Linuxbrew.

**Comment by KyleTryon** [comment]
I think we may release a patch that installs twice, we could always patch it out again. Expect that coming soon, will update.

Edit: Ok that did not work. Did someone have a link to their work around working by installing twice? Clearly some pathing issue..

**Comment by ldttmt** [comment]
We did a workaround by install aws-sam-cli twice (like @lightpriest mentioned) but it stopped working. Upon the second installation, we got this error: 
```
==> Installing aws-sam-cli from aws/tap

==> Downloading https://github.com/aws/aws-sam-cli/releases/download/v1.2.0//aws-sam-cli-1.2.0.x86_64_linux.bottle.tar.gz

Already downloaded: /home/travis/.cache/Homebrew/downloads/8bd3131572304aab77325b1bd7c22ce7efa7bcdd6b0dfc8884f68cd6788d0857--aws-sam-cli-1.2.0.x86_64_linux.bottle.tar.gz

==> Pouring aws-sam-cli-1.2.0.x86_64_linux.bottle.tar.gz

🍺  /home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.2.0: 4,143 files, 90.5MB

/home/linuxbrew/.linuxbrew/Cellar/aws-sam-cli/1.2.0/libexec/bin/python3.7: error while loading shared libraries: libpython3.7m.so.1.0: cannot open shared object file: No such file or directory
```

**Comment by btmcnellis** [comment]
I got that error today as well @ldttmt. I eventually gave up, switched the executor image to `ubuntu-1604:202007-01` rather than using whatever the default image is for this orb (in order to get a more recent version of Python easily), and used pip to install SAM instead.

I eventually got the command down to just this:

```
commands:
  sam-cli-install:
    description: Installs the AWS SAM CLI in order to build or deploy serverless applications
    steps:
      - run:
          name: Set Python version
          command: |
            pyenv versions
            pyenv global 3.8.3
      - run:
          name: Install the AWS SAM CLI
          command: |
            pip install aws-sam-cli
            sam --version
```

This is probably not the way to go in the long term, since AWS doesn't officially support installing SAM with pip, but it's better than dealing with these Homebrew issues for the second time in a week.

**Comment by kn100** [comment]
I'm not certain but here's a PR that I think addresses the issue https://github.com/CircleCI-Public/aws-sam-serverless-orb/pull/21 - I just followed what Amazon is doing here https://github.com/aws/amazon-chime-sdk-js/pull/672

If they're happy to install their own stuff using unofficial channels I don't see why CircleCI should be bothered :P

**Comment by KyleTryon** [comment]
> I'm not certain but here's a PR that I think addresses the issue #21 - I just followed what Amazon is doing here [aws/amazon-chime-sdk-js#672](https://github.com/aws/amazon-chime-sdk-js/pull/672)
> 
> If they're happy to install their own stuff using unofficial channels I don't see why CircleCI should be bothered :P

Sold.

**Comment by jfuss** [comment]
> > I'm not certain but here's a PR that I think addresses the issue #21 - I just followed what Amazon is doing here [aws/amazon-chime-sdk-js#672](https://github.com/aws/amazon-chime-sdk-js/pull/672)
> > If they're happy to install their own stuff using unofficial channels I don't see why CircleCI should be bothered :P
> 
> Sold.

We can't really stop anyone from install with pip. We highly advise against it. If you are, you should be following python best practices and install in its own venv. If you don't, you will get conflicts on dependencies (we pin various dependencies and the CLI is not meant to be installed with other tools, packages etc, which is why we prefer people not to use pip and do not document it).

Edit: I am one of the maintainers of SAM.

**Comment by DanielFrank** [comment]
> Edit: I am one of the maintainers of SAM

@jfuss Not nagging, just asking. There's two tickets in SAM homebrew Github related to this issue. Is anyone actively looking at it?

**Comment by KyleTryon** [comment]
Hey folks. Version 2.0.0 was released using Pip. We may change this out in the future but it seemed to be an appropriate fix while we try to figure out what's going on with linuxbrew.

You can check out the new version here: https://circleci.com/orbs/registry/orb/circleci/aws-sam-serverless

This also features Ubuntu 20.04 as the default machine image.

Let us know if you run into any issues!

**Comment by jfuss** [comment]
@DanielFrank We are actively looking at providing the same experience as AWS CLI does with V2. 

@KyleTryon and I had a discussion that to unblock customers this Orb can use Pip but once we release a similar solution (binary) to AWS CLI V2, we would push this Orb (and others) to install through that. 

We do not have timelines for when we will have this released but we do have engineers actively working through providing a better installation story.

**Comment by jfuss** [comment]
@KyleTryon Following back up here. We just released a new way to install with Linux. From previous discussions, I know you where interested in it for this project. We are still waiting on docs to go live but here is a post in Github for how to install: https://github.com/aws/aws-sam-cli/issues/1424#issuecomment-776982344. Hopefully that can help you reduce some risk and improve the installation process in this Orb.

**Comment by gmemstr** [comment]
@jfuss Thanks for following up! Taking a look at this myself and will bee sure to surface.

**Comment by KyleTryon** [comment]
Thank you @jfuss, this is actually excellent timing as we are working on an update right now as well.
